File: backend/app/models/seller.py
"""Seller SQLAlchemy model."""

from uuid import UUID

from sqlalchemy import Boolean, Index, String, Text, UniqueConstraint
from sqlalchemy.dialects.postgresql import UUID as PostgresUUID
from sqlalchemy.orm import Mapped, mapped_column

from app.database.base import Base, SoftDeleteMixin, TimestampMixin, UUIDMixin


class Seller(Base, UUIDMixin, TimestampMixin, SoftDeleteMixin):
    """Seller profile and business information."""

    __tablename__ = "sellers"
    __table_args__ = (
        UniqueConstraint("gst_number", name="uq_sellers_gst_number"),
        UniqueConstraint("pan_number", name="uq_sellers_pan_number"),
        UniqueConstraint("business_email", name="uq_sellers_business_email"),
        Index("ix_sellers_business_name", "business_name"),
        Index("ix_sellers_gst_number", "gst_number"),
        Index("ix_sellers_pan_number", "pan_number"),
        Index("ix_sellers_business_email", "business_email"),
        Index("ix_sellers_status", "status"),
        Index("ix_sellers_user_id", "user_id"),
    )

    user_id: Mapped[UUID] = mapped_column(PostgresUUID(as_uuid=True), nullable=False)

    business_name: Mapped[str] = mapped_column(String(255), nullable=False)
    legal_business_name: Mapped[str | None] = mapped_column(String(255), nullable=True)
    business_type: Mapped[str] = mapped_column(String(50), nullable=False)
    business_email: Mapped[str] = mapped_column(String(255), nullable=False)
    business_phone: Mapped[str] = mapped_column(String(20), nullable=False)
    gst_number: Mapped[str] = mapped_column(String(15), nullable=False)
    pan_number: Mapped[str] = mapped_column(String(10), nullable=False)
    tax_identification_number: Mapped[str | None] = mapped_column(String(50), nullable=True)
    website: Mapped[str | None] = mapped_column(String(255), nullable=True)
    logo_url: Mapped[str | None] = mapped_column(String(500), nullable=True)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)

    address_line_1: Mapped[str] = mapped_column(String(255), nullable=False)
    address_line_2: Mapped[str | None] = mapped_column(String(255), nullable=True)
    city: Mapped[str] = mapped_column(String(100), nullable=False)
    state: Mapped[str] = mapped_column(String(100), nullable=False)
    country: Mapped[str] = mapped_column(String(100), nullable=False)
    postal_code: Mapped[str] = mapped_column(String(20), nullable=False)

    account_holder_name: Mapped[str] = mapped_column(String(255), nullable=False)
    bank_name: Mapped[str] = mapped_column(String(255), nullable=False)
    account_number: Mapped[str] = mapped_column(String(34), nullable=False)
    ifsc_code: Mapped[str] = mapped_column(String(11), nullable=False)
    branch_name: Mapped[str | None] = mapped_column(String(255), nullable=True)

    default_currency: Mapped[str] = mapped_column(String(3), default="INR", nullable=False)
    notifications_enabled: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)
    order_auto_accept_enabled: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)

    is_active: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False, index=True)
    is_verified: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    status: Mapped[str] = mapped_column(String(30), default="pending", nullable=False)

File: frontend/src/types/seller.ts
import type { ApiResponse } from "@/types/common";

export type SellerStatus = "pending" | "active" | "inactive" | "suspended" | "deleted";

export type SellerBusinessType = "individual" | "proprietorship" | "partnership" | "private_limited" | "llp" | "other";

export interface Seller {
  id: string;
  user_id: string;
  business_name: string;
  legal_business_name: string | null;
  business_type: SellerBusinessType;
  business_email: string;
  business_phone: string;
  gst_number: string;
  pan_number: string;
  tax_identification_number: string | null;
  website: string | null;
  logo_url: string | null;
  description: string | null;
  address_line_1: string;
  address_line_2: string | null;
  city: string;
  state: string;
  country: string;
  postal_code: string;
  account_holder_name: string;
  bank_name: string;
  account_number: string;
  ifsc_code: string;
  branch_name: string | null;
  default_currency: string;
  notifications_enabled: boolean;
  order_auto_accept_enabled: boolean;
  is_active: boolean;
  is_verified: boolean;
  status: SellerStatus;
  created_at: string;
  updated_at: string;
  deleted_at: string | null;
  is_deleted: boolean;
}

export interface SellerListMeta {
  page: number;
  page_size: number;
  total_items: number;
  total_pages: number;
  has_next: boolean;
  has_previous: boolean;
}

export interface SellerListPayload {
  items: Seller[];
  meta: SellerListMeta;
}

export interface SellerListParams {
  page: number;
  page_size: number;
  search?: string;
  status?: SellerStatus;
  business_type?: SellerBusinessType;
  is_active?: boolean;
  is_verified?: boolean;
  city?: string;
  state?: string;
  country?: string;
  sort_by?: string;
  sort_direction?: "asc" | "desc";
}

export type SellerCreatePayload = Omit<
  Seller,
  "id" | "is_active" | "is_verified" | "status" | "created_at" | "updated_at" | "deleted_at" | "is_deleted"
>;

export type SellerUpdatePayload = Partial<Omit<SellerCreatePayload, "user_id">>;

export type SellerResponse = ApiResponse<Seller>;
export type SellerListResponse = ApiResponse<SellerListPayload>;
